### DISCUSSION TITLE
C_CanNpcCollideWithSpell

### DISCUSSION THREAD

**Discussion by Kor-Angar** [dosinabox/g2nr_unofficial_update#540] [General]
https://github.com/dosinabox/g2nr_unofficial_update/discussions/453
https://github.com/dosinabox/g2nr_unofficial_update/discussions/462

**Comment by Kor-Angar** [comment]
@dosinabox  у тебя spellType == SPL_Icebolt сгуппирован с заморозкой. лучше  SPL_Icebolt перенести к SPL_IceLance

**Reply by dosinabox** [reply]
а у SPL_Icebolt нет каких-то дополнительных эффектов?

**Comment by Kor-Angar** [comment]
вот под  эти три проверки можно было бы тоже сделать какую то общую

(C_NpcIsDown(self) || C_BodyStateContains(self,BS_SWIM) || C_BodyStateContains(self,BS_DIVE)

**Reply by dosinabox** [reply]
C_NpcIsSwimming. Обнови скрипты.

**Reply by Kor-Angar** [reply]
а  у тебя C_NpcIsSwimming используется только в  C_CanNpcCollideWithSpell  ?

**Reply by dosinabox** [reply]
не только, много где

**Comment by Kor-Angar** [comment]
надо бы добавить еще и проверку для больших монстров.

большие монстры это демоны, драконы, големы и тролли в понимание пираний :-)

**Reply by Kor-Angar** [reply]
> 		// alle grosse monster bekommen nur schaden, kein opfer zs

**Reply by dosinabox** [reply]
а мракорисы, скелеты мракорисов и огненные дьяволы - большие?
а ищейки?

**Reply by Kor-Angar** [reply]
мракорисы не большие, но кое где кое что на них не действует.

**Reply by Kor-Angar** [reply]
вот лучше отдельно сделать больших как пираньи прописали. а мракрисов и прочее добавлять доп проверкой там где это надо.

**Reply by dosinabox** [reply]
а кулак ветра должен их отбрасывать?

**Reply by Kor-Angar** [reply]
кулак ветра их не отбрасывает. а вот наносит урон или нет надо отдельно обсуждать..


там удалил. сюда скопировал

 
```D
    if(spellType == SPL_Whirlwind)
	{
		if C_NpcIsDown(self)        { return COLL_DONOTHING; };
        if C_NpcIsSwimming(self)    { return COLL_DONOTHING; };
        
        if (C_NpcIsSuperBigCreature(self)) 		        { return COLL_DONOTHING; };
        if (self.guild == GIL_WISP)                     { return COLL_DONOTHING; };
        if  (self.flags == NPC_FLAG_IMMORTAL)           { return COLL_DONOTHING; };
        if  (self.flags == GIL_SHADOWBEAST)             { return COLL_DONOTHING; };
        if  (self.flags == GIL_SHADOWBEAST_SKELETON)    { return COLL_DONOTHING; };
        if  (self.flags == GIL_SHADOWBEAST_FIRE)        { return COLL_DONOTHING; };
        //if  (self.flags == GIL_Gargoyle) { return COLL_DONOTHING; };
         
         /*
        || C_NpcIsGolem(self) || (self.guild == GIL_WISP) || C_NpcIsDemon(self) || (self.guild == GIL_TROLL) || (self.guild == GIL_DRAGON) || (self.flags == NPC_FLAG_IMMORTAL) || (self.guild == GIL_SHADOWBEAST) || (self.guild == GIL_SHADOWBEAST_SKELETON) || (self.guild == GIL_SHADOWBEAST_FIRE))
		{
			return COLL_DONOTHING;
		};
        */
		if(C_NpcIsGateGuard(self))
		{
			return COLL_APPLYDAMAGE;
		};
		return COLL_DOEVERYTHING;
	};
```

**Reply by Kor-Angar** [reply]
там удалил тут добавил

```D
	if(spellType == SPL_Geyser)
	{
		if(C_NpcIsFireCreature(self))
		{
			return COLL_APPLYDOUBLEDAMAGE;
		};
        
        if(C_NpcIsDown(self))                   { return COLL_DONOTHING; };
        if(C_NpcIsSwimming(self))                   { return COLL_DONOTHING; };
        if (C_NpcIsSuperBigCreature(self)) 		{ return COLL_DONOTHING; };
        if (self.guild == GIL_WISP)             { return COLL_DONOTHING; };
        /*
		if(C_NpcIsDown(self) || C_NpcIsSwimming(self) || C_NpcIsGolem(self) || (self.guild == GIL_WISP) || C_NpcIsDemon(self) || (self.guild == GIL_TROLL) || (self.guild == GIL_DRAGON))
		{
			return COLL_DONOTHING;
		};
        */
        
		return COLL_APPLYDAMAGE | COLL_DONTKILL;
	};
    
```

**Comment by Kor-Angar** [comment]
**SPL_WindFist**

должен ли наносить урон большим монстрам. такой вот вопрос требует ответа.

для этого надо посмотреть еще и защиты у монстров от падения.

**Reply by dosinabox** [reply]
Пусть наносит, это нормально регулируется защитой. Главное чтобы не отбрасывал.

**Comment by Kor-Angar** [comment]
**Waterfist**

надо посмотреть как нибудь и вникнуть почему у водных созданий нет COLL_APPLYHALVEDAMAGE

**Comment by Kor-Angar** [comment]
SPL_GreenTentacle

не действует на тролей, но действует на големов. присмотреться.

**Comment by Kor-Angar** [comment]
SPL_BreathOfDeath

**Reply by Kor-Angar** [reply]
неоф. обновление
			if( Npc_IsPlayer(self)			{				return COLL_APPLYHALVEDAMAGE;			};

MDK
(Hlp_GetInstanceID(self) != Hlp_GetInstanceID(hero))	{				return COLL_APPLYHALVEDAMAGE;			};


надо бы вникнуть в чём там суть

**Reply by Kor-Angar** [reply]
```D
    // ---- Todeshauch ----	
    if (spellType == SPL_BreathOfDeath)
    {
    
        if (C_NpcIsUndead(self))		{ return COLL_DONOTHING;	}; // не действует на нежить
    
        if (Npc_GetDistToNpc (other,self) >= 1000)  { return COLL_DONOTHING;	}; // дым рассеялся
        
        if (self.guild == GIL_DRAGON)   { return COLL_APPLYHALVEDAMAGE;	}; // тут наверное речь идёт про дракона нежить?? или про огненного дракона? дракон нежить проходит проверку по C_NpcIsUndead ?
        
        //if (Hlp_GetInstanceID(self) != Hlp_GetInstanceID(hero))	 {	return COLL_APPLYHALVEDAMAGE;	}; // проверка MDK опечатка?
        if (Hlp_GetInstanceID(self) == Hlp_GetInstanceID(hero))	 {	return COLL_APPLYHALVEDAMAGE;	}; // в классике всё ок. в аддоне была опечатка

        
        return COLL_DOEVERYTHING;

    };
```

**Reply by Kor-Angar** [reply]
в неоф. обновление SPL_BreathOfDeath по моему не использует глава ищущих? вроде в неоф. обновление он стал бросаться черепом?

**Reply by Kor-Angar** [reply]
> тут наверное речь идёт про дракона нежить?? или про огненного дракона? дракон нежить проходит проверку по C_NpcIsUndead ?

наверное про дракона нежить. ведь его можно убить волной смерти. тогда наверное лучше указать его явно в начале и сказать что он получает половину урона.

а огненному зачем защиту в таком случае давать? вопрос на подумать.

**Reply by dosinabox** [reply]
> в неоф. обновление SPL_BreathOfDeath по моему не использует глава ищущих? вроде в неоф. обновление он стал бросаться черепом?

не менялось

**Reply by Kor-Angar** [reply]
а в классике какой код? так?

```D
	her = Hlp_GetNpc(PC_Hero);
	if(spellType == SPL_BreathOfDeath)
	{
		if(Npc_GetDistToNpc(other,self) >= 1000)
		{
			return COLL_DONOTHING;
		};
		if((self.guild != GIL_DRAGON) && (Hlp_GetInstanceID(self) != Hlp_GetInstanceID(her)))
		{
			return COLL_DOEVERYTHING;
		}
		else
		{
			return COLL_APPLYHALVEDAMAGE;
		};
		return COLL_DONOTHING;
	};
```

**Reply by dosinabox** [reply]
> а в классике какой код? так?

да

**Reply by Kor-Angar** [reply]
в аддоне  DMT_1299_OberDementor_DI использует SPL_Cost_Skull
в классике DMT_1299_OberDementor_DI использует SPL_Deathball

тогда не понятно кто использовал дым смерти против героя :-) видимо это были какие то ранние мысли пираней которые не были воплощены в игре.

**Comment by Kor-Angar** [comment]
**SPL_Shrink**

**Reply by Kor-Angar** [reply]
собрал функции в один файл. в итоге в очередной раз коснулся Дракона. получается что условия в CanNpcCollideWithSpell не срабатывают так как уменьшение находит в функции каст, а не в magic

**Reply by Kor-Angar** [reply]
соответственно можно перенести условия из в CanNpcCollideWithSpell в Spell_Cast_Shrink

**Reply by Kor-Angar** [reply]
```D
func void Spell_Cast_Shrink()
{
	//self.attribute[ATR_MANA] -=  SPL_Cost_Shrink;
    if (C_NpcIsDown(other))                     { return; };    // условие из C_CanNpcCollideWithSpell
    if (C_BodyStateContains(other,BS_SWIM))     { return; };    // условие из C_CanNpcCollideWithSpell
    if (C_BodyStateContains(other,BS_DIVE))     { return; };    // условие из C_CanNpcCollideWithSpell
    if (other.guild == GIL_DRAGON)              { return; };    // условие из C_CanNpcCollideWithSpell

    
    if (other.flags == NPC_FLAG_IMMORTAL)       { return; };	//nicht auf Immortals
    if (C_NpcIsUndead(other)) 					{ return; };    //nicht auf Untote
	if (other.guild <= GIL_SEPERATOR_HUM)       { return; };	//nicht auf Humans
    if (other.aivar[AIV_MM_ShrinkState] != 0)   { return; };    //nur auf bisher ungeschrumpfte Monster!
   
    var string msg;
    msg = Str_Format("Spell_Cast_Shrink[%s --> %s]", self.name, other.name);
    Hlp_PrintConsole(msg);

    Npc_ClearAIQueue	(other);
    B_ClearPerceptions	(other);
    AI_StartState		(other, ZS_MagicShrink, 0, "");
	//self.aivar[AIV_SelectSpell] += 1;
};
```
